Fully document setup of a working CL development environment

edwardchalstrey1 opened the pull request:
# Does

Combines @tomalrussell changes on `feature/node16` branch to `package.json` with updated dev environment setup doc.

### TODO:
- [x] Get dev db from dump file to work
- [x] Test loading OSM polygons works
- [x] Update main README and other repo docs to avoid any duplication
- [x] Link docs from the main README more sensibly - i.e. link to dev setup and prod setup in `/docs`

# Tested with

Ubuntu 18.04.6 virtualbox vm

Closes #766

matkoniecz commented:
@edwardchalstrey1 What is the status of this PR? Are you stuck with something and requiring help?

edwardchalstrey1 commented:
> @edwardchalstrey1 What is the status of this PR? Are you stuck with something and requiring help?
> 
> Or is it ready for review?
> 
> Or just WIP backup?

@matkoniecz I think it's working - just trying to get my virtualbox NAT port forwarding to work correctly

edwardchalstrey1 commented:
@matkoniecz I can remove the `Vagrantfile` and `vm_provision.sh` if needed from this PR as they don't fix it yet

edwardchalstrey1 commented on docs/setup-dev-environment.md:
```diff
@@ -162,11 +172,55 @@ local repository, so that it can read from the `package.json` file.
 
 `cd ./colouring-london/app && npm install`
 
+## Load OpenStreetMap test polygons
```
Hi @matkoniecz @tomalrussell - I have got the application running using the changes in this PR, but currently without data loading. I think it makes sense to add in the OSM test polygons instructions here (I will modify the other docs to avoid duplication and link everything from the main README too).

Can you spot what, if anything I may have done wrong here and advise how my version of the site should look with these test buildings? Are they only in a small part of the city?

![Screenshot 2022-02-03 at 15 26 34](https://user-images.githubusercontent.com/5486164/152373362-de3e4b37-787a-4a77-800a-712c08c603c0.png)

matkoniecz replied:
https://github.com/colouring-cities/colouring-london/blob/master/etl/get_test_polygons.py#L21

Buildings should be around https://www.openstreetmap.org/#map=19/51.52450/-0.13387

edwardchalstrey1 replied:
ok thanks, I seem to have 1779 buildings in the db, but they don't show up on the map so unsure why

edwardchalstrey1 replied:
On `npm start` I've added the environment variable: `TILECACHE_PATH=../maintenance/tiles/` <- this doesn't seem to have worked either, did I do this wrong maybe?

edwardchalstrey1 replied:
also @matkoniecz do you know the answer to this question^

edwardchalstrey1 replied:
Doesn't look like files are being created in the `tilecache` dir.
Error that repeats in the console when running the node app:

```
Postgis Plugin: authentication method 10 not supported
Connection string: 'host=localhost port=5432 dbname=testlondon user=cldbadmin connect_timeout=4 fallback_application_name=mapnik'

    at renderDataSourceTile (/home/ed/colouring-london/app/colouring-london-app/src/tiles/renderers/renderDataSourceTile.ts:34:1)
    at renderBuildingTile (/home/ed/colouring-london/app/build/webpack:/colouring-london-app/src/tiles/rendererDefinition.ts:47:1)
    at stitchOrRenderBuildingTile (/home/ed/colouring-london/app/build/webpack:/colouring-london-app/src/tiles/rendererDefinition.ts:57:1)
    at getTileWithCaching (/home/ed/colouring-london/app/build/webpack:/colouring-london-app/src/tiles/renderers/getTileWithCaching.ts:7:1)
    at /home/ed/colouring-london/app/build/webpack:/colouring-london-app/src/tiles/tileserver.ts:20:1
```

The suggestion I've found online is that `libpq`needs to be updated, however, when I try installing the following I get:

```
libpq-dev is already the newest version (14.2-1.pgdg18.04+1).
```

&

```
libpq5 is already the newest version (14.2-1.pgdg18.04+1).
```

tomalrussell replied:
It is likely that `node-mapnik` was build with a different `libpq`. It looks like this issue is reported at https://github.com/mapnik/mapnik/issues/4283 

Can you try the workaround of setting postgres to store passwords with `md5`? [Postgres docs](https://www.postgresql.org/docs/14/auth-password.html)

(Otherwise could look into building node-mapnik - see for example slightly old notes in https://github.com/mapnik/mapnik-support/issues/110)

edwardchalstrey1 replied:
Ok, for now I have it working with Postgres-12, I can look at that issue properly at a later date when working on #531

matkoniecz commented on docs/setup-dev-environment.md:
````diff
@@ -162,11 +172,55 @@ local repository, so that it can read from the `package.json` file.
+psql -d <colouringlondondb> < ../migrations/003.index-buildings.up.sql
+```
+
+## Finally
````
https://github.com/colouring-cities/colouring-london/blob/master/migrations/003.index-buildings.up.sql looks like a pure performance improvement

edwardchalstrey1 replied:
Sorry this comment is referring to what I added below ## Finally

edwardchalstrey1 commented:
hmm, frustratingly my Docker version also fails at the `npm install` step like in Vagrant - but it did work with @tomalrussell updates to `package.json` when I ran the setup steps on my virtualbox vm manually.

I'll need to investigate further what the difference is

matkoniecz commented on Dockerfile:
```diff
@@ -0,0 +1,34 @@
+
+RUN apt-get install -y build-essential wget
+RUN apt-get install -y python3 python3-pip python3-dev python3-venv
+RUN apt install parallel -y
```
```suggestion
RUN apt-get install parallel -y
```

matkoniecz commented on Dockerfile:
```diff
@@ -0,0 +1,34 @@
+COPY etl /colouring-london/etl
+
+ENV NODE_VERSION=16.13.2
+RUN apt install -y curl
```
```suggestion
RUN apt-get install -y curl
```

This prevents some warning messages from appearing in logs

matkoniecz commented:
On just running standard docker compose I end with

```
colouring-london    | secret option required for sessions
colouring-london    |     at session (/colouring-london/app/node_modules/express-session/index.js:200:12)
colouring-london    |     at Layer.handle [as handle_request] (/colouring-london/app/node_modules/express/lib/router/layer.js:95:5)
colouring-london    |     at trim_prefix (/colouring-london/app/node_modules/express/lib/router/index.js:323:13)
colouring-london    |     at /colouring-london/app/node_modules/express/lib/router/index.js:284:7
colouring-london    |     at Function.process_params (/colouring-london/app/node_modules/express/lib/router/index.js:341:12)
colouring-london    |     at next (/colouring-london/app/node_modules/express/lib/router/index.js:275:10)
colouring-london    |     at SendStream.error (/colouring-london/app/node_modules/serve-static/index.js:121:7)
colouring-london    |     at SendStream.emit (node:events:390:28)
colouring-london    |     at SendStream.error (/colouring-london/app/node_modules/send/index.js:270:17)
colouring-london    |     at SendStream.onStatError (/colouring-london/app/node_modules/send/index.js:421:12)
```

in logs and 

```
secret option required for sessions
    at session (/colouring-london/app/node_modules/express-session/index.js:200:12)
    at Layer.handle [as handle_request] (/colouring-london/app/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/colouring-london/app/node_modules/express/lib/router/index.js:323:13)
    at /colouring-london/app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/colouring-london/app/node_modules/express/lib/router/index.js:341:12)
    at next (/colouring-london/app/node_modules/express/lib/router/index.js:275:10)
    at SendStream.error (/colouring-london/app/node_modules/serve-static/index.js:121:7)
    at SendStream.emit (node:events:390:28)
    at SendStream.error (/colouring-london/app/node_modules/send/index.js:270:17)
    at SendStream.onStatError (/colouring-london/app/node_modules/send/index.js:421:12)
```

on http://127.0.0.1:8080/ page.

Should I do some extra steps?

tomalrussell commented on docker-compose.yml:
```diff
@@ -0,0 +1,46 @@
+      POSTGRES_USER: dockeruser
+      POSTGRES_PASSWORD: postgres
+      POSTGRES_DB: colouring-london-db
+      APP_COOKIE_SECRET: 123456
```
`APP_COOKIE_SECRET` should be an environment variable for the `colouring-london` node app container - this is probably causing the error @matkoniecz sees.

edwardchalstrey1 replied:
fantastic, thanks Tom - one step closer 😄 :
![Screenshot 2022-02-09 at 09 50 39](https://user-images.githubusercontent.com/5486164/153171518-74e4f2c3-1687-45ff-b13f-625a7b9d7ead.png)

edwardchalstrey1 commented:
> On just running standard docker compose I end with
> 
> ```
> colouring-london    | secret option required for sessions
> colouring-london    |     at session (/colouring-london/app/node_modules/express-session/index.js:200:12)
> colouring-london    |     at Layer.handle [as handle_request] (/colouring-london/app/node_modules/express/lib/router/layer.js:95:5)
> colouring-london    |     at trim_prefix (/colouring-london/app/node_modules/express/lib/router/index.js:323:13)
> colouring-london    |     at /colouring-london/app/node_modules/express/lib/router/index.js:284:7
> colouring-london    |     at Function.process_params (/colouring-london/app/node_modules/express/lib/router/index.js:341:12)
> colouring-london    |     at next (/colouring-london/app/node_modules/express/lib/router/index.js:275:10)
> colouring-london    |     at SendStream.error (/colouring-london/app/node_modules/serve-static/index.js:121:7)
> colouring-london    |     at SendStream.emit (node:events:390:28)
> colouring-london    |     at SendStream.error (/colouring-london/app/node_modules/send/index.js:270:17)
> colouring-london    |     at SendStream.onStatError (/colouring-london/app/node_modules/send/index.js:421:12)
> ```
> 
> in logs and
> 
> ```
> secret option required for sessions
>     at session (/colouring-london/app/node_modules/express-session/index.js:200:12)
>     at Layer.handle [as handle_request] (/colouring-london/app/node_modules/express/lib/router/layer.js:95:5)
>     at trim_prefix (/colouring-london/app/node_modules/express/lib/router/index.js:323:13)
>     at /colouring-london/app/node_modules/express/lib/router/index.js:284:7
>     at Function.process_params (/colouring-london/app/node_modules/express/lib/router/index.js:341:12)
>     at next (/colouring-london/app/node_modules/express/lib/router/index.js:275:10)
>     at SendStream.error (/colouring-london/app/node_modules/serve-static/index.js:121:7)
>     at SendStream.emit (node:events:390:28)
>     at SendStream.error (/colouring-london/app/node_modules/send/index.js:270:17)
>     at SendStream.onStatError (/colouring-london/app/node_modules/send/index.js:421:12)
> ```
> 
> on http://127.0.0.1:8080/ page.
> 
> Should I do some extra steps?

@matkoniecz I'm still working on this, looks like Tom's suggestion above has helped though

edwardchalstrey1 commented on docs/setup-dev-environment.md:
```diff
@@ -93,34 +101,91 @@ Restart postgres to pick up config changes.
+## Loading the data
+
+<details>
+<summary> With a database dump </summary>
```
@tomalrussell @matkoniecz I was wondering if you both did something similar to this when restoring the database from a dump file in a dev environment (virtualbox Ubuntu VM)? Essentially I ran through all the prior postgres steps in this doc ([see rendered md](https://github.com/colouring-cities/colouring-london/blob/docs-vagrant/docs/setup-dev-environment.md#loading-the-data)) up to this point, with the `<username>` and `<pgpassword>` set the same as the username and pass in the staging db, where I got the dump file (I'm unsure whether choosing the same username and password was neccesarry).

However the real problem (I think) seems to be that the VM quickly runs out of space, I set mine to have 10GB of storage for a 4GB dumpfile <- I thought this would be more than enough as one might expect the db to require less space than a file.

Not entirely sure that is the problem, I'm getting many repeats of this output: `invalid command \N` on running `psql -d <colouringlondondb> -U <username> -h localhost < <dumpfile>`, after it does some stuff, but it's hard to see which command failed as I can't scroll all the way up. Running it again results in:
```
^CERROR:  could not extend file "base/16385/17567": No space left on device
HINT:  Check free disk space.
```
hence I assume the space issue! Any help would be welcome. :)

edwardchalstrey1 replied:
Ok, I tried everything again with a 50GB storage VM (dynamically allocated).

I tried again to create the db from the dump file I downloaded from staging server and mounted:

```
psql -d colouringlondondb -U cldbadmin -h localhost < db-dumps/colouringlondondb_staging_2022-02-14.dump
```

which results in the following error with `invalid command \N` repeated many times:

```
invalid command \N
invalid command \N
invalid command \.
ERROR:  syntax error at or near "3411839"
LINE 1: 3411839 osgb5000005210563591
        ^
invalid command \.
ERROR:  syntax error at or near "1"
LINE 1: 1 2019-10-03 05:33:00+00 /var/www/colouring-london/downloads...
        ^
invalid command \.
Killed
```

Perhaps I will try again with a different dump file.

edwardchalstrey1 replied:
Or should I have created the db from the dumpfile before running the migrations files, perhaps that is the problem, I will check

edwardchalstrey1 replied:
aha creating a fresh db with the dumpfile seems to have worked, although looks like I need to also add an extra user:

```
ERROR:  role "clwebapp" does not exist
```

tomalrussell replied:
Yes, should be easiest to restore the dump to an empty database, but with the expected postgres users/roles created.

edwardchalstrey1 replied:
Thanks @tomalrussell looks like I have that working, now facing the same issue with `libpq` as in the above thread, I'll try your idea!

edwardchalstrey1 commented:
Hi @tomalrussell @matkoniecz - I have managed to set up databases in a development environment, both from the test OSM polygons and from a dump of the database from the staging server. I've gotten both working with the application and updated the documentation with the steps that worked for me. If all looks good to you, we can merge this and I will respond to Geoffroy on #766 

Updated guide: https://github.com/colouring-cities/colouring-london/blob/docs-vagrant/docs/setup-dev-environment.md
Updated README that shows people where to look: https://github.com/colouring-cities/colouring-london/tree/docs-vagrant#setup-and-run

Also @tomalrussell I ended up just using Postgres-12 for now, there is a separate issue on the problem with Postgres-14 open, I can look at that later, but doesn't seem like a priority.

Also, ignore the branch name, I have removed any Vagrant updates to a separate branch - however I'll aim to continue work on my Docker implementation next

matkoniecz commented on docs/setup-dev-environment.md:
````diff
@@ -57,129 +130,282 @@ EOF
+service postgresql restart
+```
+
+<!-- Change the password encryption to md5. You will need to edit the postgres config file manually, for example you could use nano, then search for `password_encryption` and change it from `scram-sha-256` to `md5` (making sure you un-comment this line).
````
is it intentional that it is added but as commented out?

edwardchalstrey1 replied:
Thanks for spotting this, I can remove, it was a thing I was trying to make postgres-14 work with mapnik.

matkoniecz commented on docs/setup-dev-environment.md:
```diff
@@ -57,129 +130,282 @@ EOF
 
-The site can then be viewed on http://localhost:8080. The `app_dev` mapping is used in
-development by Razzle which rebuilds and serves client side assets on the fly.
+The site can then be viewed on http://localhost:8080.
```
```suggestion
The site can then be viewed on http://localhost:8080 on the host computer.
```

clarify where it can be seen (not tested, based on config listed above)

matkoniecz commented on docs/setup-dev-environment.md:
```diff
@@ -1,43 +1,116 @@
-Colouring London. This guide assumes you already have Ubuntu 20.04 server installed, typically
-installed in a virtual environment such a Virtual Box and are able to SSH into your Ubuntu
-installation for convenience.
+This document is intended to guide you through setting up a local development environment for the Colouring London application. This guide assumes you already have either already have access to an machine with Ubuntu 18.04 or 20.04 installed, or can use VirtualBox to set up an Ubuntu virtual machine as below:
```
Is it also supposed to work on 20.04? Or at least beneficial to test with 20.04?

edwardchalstrey1 replied:
I tested on 18.04 simply because I forgot to apply the update to 20 when I created my VM in Virtualbox - however I assume it works on 20 if that is what you have used? I could test this further to be sure, what do you think?

matkoniecz replied:
I plan it on testing anyway, so I can do this and let you know if anything breaks (and retest on 18.04 in such case)

edwardchalstrey1 replied:
ok great thanks

matkoniecz commented on docs/setup-dev-environment.md:
```diff
@@ -1,43 +1,116 @@
+
+To run the commands in the rest of this setup guide, either `ssh` into the VirtualBox environment or open the terminal within the Ubuntu GUI.
+
+If you wish to `ssh`, you will first need to open the terminal in Ubuntu and run the following.
```
```suggestion
If you wish to `ssh`, you will first need to open the terminal in Ubuntu and run the following:
```

very minor stylistic change (proposed just here for now to check is it welcomed)

edwardchalstrey1 replied:
ok yes, I can change throughout to make consistent

matkoniecz commented on docs/setup-dev-environment.md:
```diff
@@ -1,43 +1,116 @@
+### :red_circle: Installing PostgreSQL
 
-Now we install some essential tools.
+Set the postgres repo for apt.
```
maybe link here the source of this specific install instructions? To allow easier confirmation that it is the correct one and to make future updates easier?

matkoniecz commented:
@edwardchalstrey1 I have some very minor comments (based on reading diff), my main question is it is supposed to work also with 20.04? (or is it at least worth testing with that OS?)

edwardchalstrey1 commented:
> @edwardchalstrey1 I have some very minor comments (based on reading diff), my main question is it is supposed to work also with 20.04? (or is it at least worth testing with that OS?)

sure ok then, see my comment above, I guess I can check - in fact it would be good to run through the steps once more anyway

matkoniecz commented on docs/setup-dev-environment.md:
````diff
@@ -1,43 +1,116 @@
+Now clone the `colouring-london` codebase. 
+
+```bash
+git clone https://github.com/colouring-london/colouring-london.git
````
```suggestion
cd ~ && git clone https://github.com/colouring-london/colouring-london.git
```

matkoniecz replied:
Make easier to follow instructions (cloning the repo into the home directory)

matkoniecz commented on docs/setup-dev-environment.md:
````diff
@@ -57,129 +130,282 @@ EOF
+
+First install prerequisites.
+```bash
+sudo apt-get install parallel
````
```suggestion
sudo apt-get install -y parallel
```

matkoniecz commented on docs/setup-dev-environment.md:
````diff
@@ -48,138 +121,285 @@ rm node-$NODE_VERSION-$DISTRO.tar.xz
 
-`service postgresql start`
+```bash
+service postgresql start
````
```suggestion
sudo service postgresql start
```

sudo is needed according to my test

matkoniecz commented on docs/setup-dev-environment.md:
````diff
@@ -48,138 +121,285 @@ rm node-$NODE_VERSION-$DISTRO.tar.xz
 
-`service postgresql restart`
+```bash
+service postgresql restart
````
```suggestion
sudo service postgresql restart
```

matkoniecz commented:
I will need to recheck it ( I tried to make it more reproducible via https://github.com/matkoniecz/turing_redo but run into some minor issues mentioned above, but that script is not fully working ). But it seems that 20.04 dies on mapnik.

```
npm install
npm ERR! code 1
npm ERR! path /home/virtual/colouring-london/app/node_modules/mapnik
npm ERR! command failed
npm ERR! command sh -c node-gyp rebuild
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@8.4.1
npm ERR! gyp info using node@16.13.2 | linux | x64
npm ERR! gyp info find Python using Python version 3.8.10 found at "/usr/bin/python3"
npm ERR! gyp http GET https://nodejs.org/download/release/v16.13.2/node-v16.13.2-headers.tar.gz
npm ERR! gyp http 200 https://nodejs.org/download/release/v16.13.2/node-v16.13.2-headers.tar.gz
npm ERR! gyp http GET https://nodejs.org/download/release/v16.13.2/SHASUMS256.txt
npm ERR! gyp http 200 https://nodejs.org/download/release/v16.13.2/SHASUMS256.txt
npm ERR! gyp info spawn /usr/bin/python3
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   '/usr/local/lib/node/node-v16.13.2/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'make',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/virtual/colouring-london/app/node_modules/mapnik/build/config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/virtual/colouring-london/app/node_modules/mapnik/common.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/usr/local/lib/node/node-v16.13.2/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/virtual/.cache/node-gyp/16.13.2/include/node/common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=/home/virtual/.cache/node-gyp/16.13.2',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/usr/local/lib/node/node-v16.13.2/lib/node_modules/npm/node_modules/node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=/home/virtual/.cache/node-gyp/16.13.2/<(target_arch)/node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=/home/virtual/colouring-london/app/node_modules/mapnik',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! gyp: Undefined variable module_name in binding.gyp while trying to load binding.gyp
npm ERR! gyp ERR! configure error 
npm ERR! gyp ERR! stack Error: `gyp` failed with exit code: 1
npm ERR! gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node/node-v16.13.2/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:259:16)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:390:28)
npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)
npm ERR! gyp ERR! System Linux 5.11.0-43-generic
npm ERR! gyp ERR! command "/usr/local/lib/node/node-v16.13.2/bin/node" "/usr/local/lib/node/node-v16.13.2/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd /home/virtual/colouring-london/app/node_modules/mapnik
npm ERR! gyp ERR! node -v v16.13.2
npm ERR! gyp ERR! node-gyp -v v8.4.1
npm ERR! gyp ERR! not ok

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/virtual/.npm/_logs/2022-02-24T16_23_55_638Z-debug-0.log
```

matkoniecz commented on docs/setup-dev-environment.md:
````diff
@@ -48,138 +121,285 @@ rm node-$NODE_VERSION-$DISTRO.tar.xz
-To test the app user's connection to the database, you could run `psql` interactively:
+Run `psql` interactively.
 
+```bash
+psql
 ```
-psql -d <colouringlondondb> -U <username> -h localhost
+
+In `psql`, necessary postgres extensions.
+
+```bash
+create extension postgis;
+create extension pgcrypto;
+create extension pg_trgm;
 ```
 
-Create the necessary postgres extensions.
+Then quit `psql` by typing `\q` and hitting return.
````
what about replacing it with

```
psql -d <colouringlondondb> -c "create extension postgis;"
psql -d <colouringlondondb> -c "create extension pgcrypto;"
psql -d <colouringlondondb> -c "create extension pg_trgm;"
```

?

edwardchalstrey1 replied:
I think this makes sense, I had a reason why I changed it, however I think that reason is no longer relevant

matkoniecz replied:
I can confirm that it works fine.

matkoniecz commented:
18.04 install also failed on mapnik (the same as mu previous attempts of my own work on this)
```
npm ERR! code 1
npm ERR! path /home/virtual/colouring-london/app/node_modules/mapnik
npm ERR! command failed
npm ERR! command sh -c node-gyp rebuild
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@8.4.1
npm ERR! gyp info using node@16.13.2 | linux | x64
npm ERR! gyp info find Python using Python version 3.6.9 found at "/usr/bin/python3"
npm ERR! gyp http GET https://nodejs.org/download/release/v16.13.2/node-v16.13.2-headers.tar.gz
npm ERR! gyp http 200 https://nodejs.org/download/release/v16.13.2/node-v16.13.2-headers.tar.gz
npm ERR! gyp http GET https://nodejs.org/download/release/v16.13.2/SHASUMS256.txt
npm ERR! gyp http 200 https://nodejs.org/download/release/v16.13.2/SHASUMS256.txt
npm ERR! gyp info spawn /usr/bin/python3
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   '/usr/local/lib/node/node-v16.13.2/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'make',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/virtual/colouring-london/app/node_modules/mapnik/build/config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/virtual/colouring-london/app/node_modules/mapnik/common.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/usr/local/lib/node/node-v16.13.2/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/virtual/.cache/node-gyp/16.13.2/include/node/common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=/home/virtual/.cache/node-gyp/16.13.2',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/usr/local/lib/node/node-v16.13.2/lib/node_modules/npm/node_modules/node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=/home/virtual/.cache/node-gyp/16.13.2/<(target_arch)/node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=/home/virtual/colouring-london/app/node_modules/mapnik',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! gyp: Undefined variable module_name in binding.gyp while trying to load binding.gyp
npm ERR! gyp ERR! configure error 
npm ERR! gyp ERR! stack Error: `gyp` failed with exit code: 1
npm ERR! gyp ERR! stack     at ChildProcess.onCpExit (/usr/local/lib/node/node-v16.13.2/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:259:16)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:390:28)
npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)
npm ERR! gyp ERR! System Linux 5.4.0-100-generic
npm ERR! gyp ERR! command "/usr/local/lib/node/node-v16.13.2/bin/node" "/usr/local/lib/node/node-v16.13.2/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd /home/virtual/colouring-london/app/node_modules/mapnik
npm ERR! gyp ERR! node -v v16.13.2
npm ERR! gyp ERR! node-gyp -v v8.4.1
npm ERR! gyp ERR! not ok

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/virtual/.npm/_logs/2022-02-25T08_59_20_072Z-debug-0.log

```

matkoniecz reviewed (commented):
I have not succeeded in using the instructions, but I attempted to make a reproducible script.

Changes here are based on attempts to make code easier to rerun and not require - for example - being in a specific folder when executing commands.

edwardchalstrey1 commented:
> https://github.com/matkoniecz/turing_redo

@matkoniecz I think this might be resolved if you check out this branch - at the moment in your script it will use the master branch and that does not have some changes to the npm packages

edwardchalstrey1 commented on docs/setup-dev-environment.md:
````diff
@@ -1,43 +1,116 @@
+Now install some essential tools.
+
+```bash
+sudo apt-get install -y build-essential git vim-nox wget curl
````
I'm not sure either!

tomalrussell replied:
This is only included in case of wanting to edit e.g. config files within the vm. Not strictly necessary for setup.

matkoniecz replied:
I think that I would remove it, as people should be able to figure how to install it and at least this setup is not including modifications to text files.

edwardchalstrey1 commented on docs/setup-dev-environment.md:
````diff
@@ -48,138 +121,285 @@ rm node-$NODE_VERSION-$DISTRO.tar.xz
 
-`echo "host    all             all             all                     md5" | sudo tee --append /etc/postgresql/10/main/pg_hba.conf > /dev/null`
+```bash
+echo "host    all             all             all                     md5" | sudo tee --append /etc/postgresql/12/main/pg_hba.conf > /dev/null
````
Could be useful to include, but I tried to make it clear in the README that these are the instructions for a development environment, so perhaps it's ok? I plan to update the [equivalent production server doc ](https://github.com/colouring-cities/colouring-london/blob/master/docs/setup-production-environment.md) later, but I think that can be a separate PR. Fell free to add a suggested commit with the disclaimer here if you think it's needed though :)

matkoniecz commented:
> @matkoniecz I think this might be resolved if you check out this branch

thanks for spotting! I was getting a bit confused what is going on

matkoniecz commented:
> Also, this is not a strong opinion so feel free to disagree, but I think these steps work better as documentation than a script that runs everything

Definitely! I created it to test that this steps are working - so I can deploy new VM and rerun script and see what is going wrong. Rather than copy paste each step over and over again.

And to test that existing documentation can be easily used to create such script (that is why I proposed some changes like avoiding `su root`). It will also make easier to test is `vim-nox` actually needed.

It also allowed easier tracking what is going wrong in my case (missing checkout).

I am definitely not proposing to replace what you created with it and lose all documentation.

(and then rerun in on 20.04)

matkoniecz commented on docs/setup-dev-environment.md:
````diff
@@ -48,138 +121,285 @@ rm node-$NODE_VERSION-$DISTRO.tar.xz
+
+### :arrow_forward: Configuring Node.js
 
+Now upgrade the npm package manager to the most recent release with global privileges. This needs to be performed as root user, so it is necessary to export the node variables to the root user profile. Don't forget to exit from root at the end.
+
+```bash
+sudo su root
+export NODEJS_HOME=/usr/local/lib/node/node-v16.13.2/bin/
+export PATH=$NODEJS_HOME:$PATH
+npm install -g npm@latest
+exit
````
I just had first successful run and it was setup with this, I think it is OK.

matkoniecz reviewed (changes requested):
I have some suggestions about improvements ( https://github.com/colouring-cities/colouring-london/pull/776#pullrequestreview-892757700 https://github.com/colouring-cities/colouring-london/pull/776#discussion_r814585411 https://github.com/colouring-cities/colouring-london/pull/776#discussion_r814470902 ) but none of that is a blocker.

I tested on 18.04 and it appears to be working (I have not tested restoring database dump or made thorough testing, everything appears to be fine).

On 20.04 it is able to open web page but import of test polygons is failing (I have not tested restoring database dump).

See https://gist.github.com/matkoniecz/36df073591b102247bb72ea3916ecb03 for log of a failure.

Adding `sudo apt-get install -y --quiet libgdal-dev` is fixing this specific error, but bumping [psycopg2 in requirements.txt](https://github.com/colouring-cities/colouring-london/pull/762/files#diff-c0cce172960ea9a6cc3db87bd46da4d8c9e397d204101303ff8df2cbbb850caa) seems needed due to https://github.com/psycopg/psycopg2/issues/854

Not sure is further work needed for 20.04.

I think it may be worth considering to use `apt-get install -y --quiet` to reduce flood of text as packages are installed.

matkoniecz commented on README.md:
```diff
@@ -26,26 +26,19 @@ for download under a liberal open data license
+
+You can try out the Colouring London application by setting up your own development environment, which includes the option to load test data from OpenStreetMaps (OSM). See [docs/setup-dev-environment](docs/setup-dev-environment.md).
+  
+_Last updated 24th February 2022_
```
```suggestion
_Last updated March 2022_
```
maybe reduced accuracy wold be better

matkoniecz commented on README.md:
```diff
@@ -26,26 +26,19 @@ for download under a liberal open data license
+
+We also have documentation on setting up a production environment here: [docs/setup-production-environment](docs/setup-production-environment.md).
+  
+_Last updated 24th December 2021_
```
```suggestion
_Last updated December 2021_
```
